### PR TITLE
autotune target_bits example for llama recipe

### PR DESCRIPTION
**Pull request by xin3he**
### **PR Type**
Enhancement

___

### **Description**
- Add `target_bits` as float in AutoRoundConfig

- Remove redundant `params_list` in multiple classes

- Introduce `non_tunable_params` in TorchBaseConfig

- Add `output_dir` to AutoRoundConfig

___

### Diagram Walkthrough

```mermaid
flowchart LR
  A["Add target_bits as float"] -- "in AutoRoundConfig" --> B["Remove redundant params_list"]
  B -- "Introduce non_tunable_params" --> C["Add output_dir"]
```

<details> <summary><h3> File Walkthrough</h3></summary>

<table><thead><tr><th></th><th align="left">Relevant files</th></tr></thead><tbody><tr><td><strong>Enhancement</strong></td><td><details><summary>1 files</summary><table>
<tr>
  <td><strong>config.py</strong><dd><code>Modify AutoRoundConfig and clean up config classes</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/intel/neural-compressor/pull/2344/files#diff-8b51e8070f15664699f40942565d708f9ee52a4e008ee6a8ec6d27af3dfe98cc">+11/-167</a></td>

</tr>
</table></details></td></tr><tr><td><strong>Additional files</strong></td><td><details><summary>11 files</summary><table>
<tr>
  <td><strong>README.md</strong></td>
  <td><a href="https://github.com/intel/neural-compressor/pull/2344/files#diff-0ef988d682f8e13111fc116f51307baebd0fcc37adf179f27166ba07eef4f6b6">+217/-0</a>&nbsp; </td>

</tr>

<tr>
  <td><strong>quantize.py</strong></td>
  <td><a href="https://github.com/intel/neural-compressor/pull/2344/files#diff-603eebf6e4814246013e06ab18a5c53880a8e6575c5446eedc71d5cb6ee73037">+260/-0</a>&nbsp; </td>

</tr>

<tr>
  <td><strong>requirements.txt</strong></td>
  <td><a href="https://github.com/intel/neural-compressor/pull/2344/files#diff-e24c61bada10c72ceb8b68cbeaf1a45c48430ceb49994b44b354011108017788">[link]</a>&nbsp; &nbsp; </td>

</tr>

<tr>
  <td><strong>README.md</strong></td>
  <td><a href="https://github.com/intel/neural-compressor/pull/2344/files#diff-0b6b7120d1feb69755caceccef86358736ac7d269257cfc483a9abb7c65e327a">+0/-125</a>&nbsp; </td>

</tr>

<tr>
  <td><strong>quantize.py</strong></td>
  <td><a href="https://github.com/intel/neural-compressor/pull/2344/files#diff-9c975404437241ea14fb46a17d0f2bb2a17295520fe0c03dad41e043f337a49a">+0/-261</a>&nbsp; </td>

</tr>

<tr>
  <td><strong>Meta-Llama-3.1-8B-Instruct_7bits.json</strong></td>
  <td><a href="https://github.com/intel/neural-compressor/pull/2344/files#diff-19c75ab346a41401ceb1b0fd1583bb4996e62eebd3ff9fe3c8aa932c2d3c9697">+0/-2242</a></td>

</tr>

<tr>
  <td><strong>Meta-Llama-3.3-70B-Instruct_5bits.json</strong></td>
  <td><a href="https://github.com/intel/neural-compressor/pull/2344/files#diff-31581dfc48abfb98107b40139574b44992b2c3fb4ac17a710949e4dd348283da">+0/-5602</a></td>

</tr>

<tr>
  <td><strong>run_hf_inf.py</strong></td>
  <td><a href="https://github.com/intel/neural-compressor/pull/2344/files#diff-4db844ab1dca9b6ea747ea021b013d89116c20561faf0fdb279ba5f907ba368d">+0/-29</a>&nbsp; &nbsp; </td>

</tr>

<tr>
  <td><strong>base_config.py</strong></td>
  <td><a href="https://github.com/intel/neural-compressor/pull/2344/files#diff-1e3238bfa826bf9c58cde91593fbc14b69c272d8856cbc16e3a3202ce7c18e53">+2/-1</a>&nbsp; &nbsp; &nbsp; </td>

</tr>

<tr>
  <td><strong>autoround.py</strong></td>
  <td><a href="https://github.com/intel/neural-compressor/pull/2344/files#diff-31f8edc7f1012e3bea2bf4839137ce63b1d32b03c65eec5a8334b31dcf4c0ef5">+16/-1</a>&nbsp; &nbsp; </td>

</tr>

<tr>
  <td><strong>auto_accelerator.py</strong></td>
  <td><a href="https://github.com/intel/neural-compressor/pull/2344/files#diff-d47c7e65cf840b709dfbe48d01c569f3a2d4dbd6f690559c8af1dd1d1155286f">+2/-1</a>&nbsp; &nbsp; &nbsp; </td>

</tr>
</table></details></td></tr></tr></tbody></table>

</details>

___

### CODE REVIEW

**Comment by PRAgent4INC** [comment]
## PR Reviewer Guide 🔍

Here are some key observations to aid the review process:

<table>
<tr><td>⏱️&nbsp;<strong>Estimated effort to review</strong>: 4 🔵🔵🔵🔵⚪</td></tr>
<tr><td>🧪&nbsp;<strong>No relevant tests</strong></td></tr>
<tr><td>🔒&nbsp;<strong>No security concerns identified</strong></td></tr>
<tr><td>⚡&nbsp;<strong>Recommended focus areas for review</strong><br><br>

<details><summary><a href='https://github.com/intel/neural-compressor/pull/2344/files#diff-31f8edc7f1012e3bea2bf4839137ce63b1d32b03c65eec5a8334b31dcf4c0ef5R386-R387'><strong>Exception Handling</strong></a>

The `except` block in the `convert` method is too broad. It catches all exceptions without any handling, which can hide errors and make debugging difficult.
</summary>

```python
pass

```

</details>

<details><summary><a href='https://github.com/intel/neural-compressor/pull/2344/files#diff-31f8edc7f1012e3bea2bf4839137ce63b1d32b03c65eec5a8334b31dcf4c0ef5R379-R381'><strong>Resource Management</strong></a>

The `del model` statement and `self.accelerator.empty_cache()` call should be inside a `finally` block to ensure they are executed even if an exception occurs.
</summary>

```python
del model
self.accelerator.empty_cache()
logger.info("Quantization is done, reloading model from saved directory...")
```

</details>

<details><summary><a href='https://github.com/intel/neural-compressor/pull/2344/files#diff-d47c7e65cf840b709dfbe48d01c569f3a2d4dbd6f690559c8af1dd1d1155286fR210-R210'><strong>Garbage Collection</strong></a>

Calling `gc.collect()` in the `empty_cache` method might not be necessary and could impact performance. Consider removing it unless profiling shows a significant benefit.
</summary>

```python
gc.collect()
```

</details>

</td></tr>
</table>

**Comment by PRAgent4INC** [comment]
## PR Code Suggestions ✨

File: neural_compressor/common/base_config.py [hunk 1]
```diff
@@ -531,7 +532,7 @@ def expand(self) -> List[BaseConfig]:
             # Assign the options to the `TuningParam` instance
             param_val = getattr(config, tuning_param.name)
             if param_val is not None:
-                if tuning_param.is_tunable(param_val):
+                if param not in self.non_tunable_params and tuning_param.is_tunable(param_val):
```
[the remark this answers is not in the document]
Comment (yiliu30) [reply neural_compressor/common/base_config.py]: make sense, please help clarify.

File: neural_compressor/torch/algorithms/weight_only/autoround.py [hunk 2]
```diff
@@ -370,6 +373,18 @@ def convert(self, model: torch.nn.Module, *args, **kwargs):
...
+        if self.export_format in ["auto_round", "llm_compressor"]:
+            # the directly returned model is QuantLinear, which is used for packing.
+            try:
+                del model
+                self.accelerator.empty_cache()
+                logger.info("Quantization is done, reloading model from saved directory...")
+                import transformers  # pylint: disable=E0401
+
+                model = transformers.AutoModelForCausalLM.from_pretrained(self.output_dir)
+            except:
```
Comment (yiliu30) [inline neural_compressor/torch/algorithms/weight_only/autoround.py:383]: Please raise some msg to the user.

File: examples/pytorch/nlp/huggingface_models/language-modeling/quantization/auto_round/llama3/README.md [hunk 3]
```diff
@@ -0,0 +1,229 @@
...
+## Inference
+
+### MXFP4 / MXFP8
+
+- Both pure MXFP4/MXFP8 and mix-precision model generated by target bits are supported.
+
+#### Prerequisite
+
+```bash
+# Install the forked vLLM
```
Comment (chensuyue) [inline examples/pytorch/nlp/huggingface_models/language-modeling/quantization/auto_round/llama3/README.md:176]: git clone -b fused-moe-ar --single-branch --quiet https://github.com/yiliu30/vllm-fork.git && cd vllm-fork
VLLM_USE_PRECOMPILED=1 pip install -e .

File: neural_compressor/torch/algorithms/weight_only/autoround.py [hunk 4]
```diff
@@ -349,6 +351,7 @@ def convert(self, model: torch.nn.Module, *args, **kwargs):
             enable_norm_bias_tuning=self.enable_norm_bias_tuning,
             truncation=self.truncation,
             enable_torch_compile=self.enable_torch_compile,
+            # TODO: AutoRound is using layer_config to quantize lm_head, remove it.
```
Comment (Kaihui-intel) [inline neural_compressor/torch/algorithms/weight_only/autoround.py]: https://github.com/intel/auto-round/blob/e57555881a67260f60dd57b0717a9987882bb8d7/auto_round/compressors/base.py#L279 
`quant_lm_head` is used in AutoRound

File: neural_compressor/torch/algorithms/weight_only/autoround.py [hunk 5]
```diff
@@ -370,6 +373,18 @@ def convert(self, model: torch.nn.Module, *args, **kwargs):
 
         dump_model_op_stats(rounder.layer_config)
 
+        if self.export_format in ["auto_round", "llm_compressor"]:
+            # the directly returned model is QuantLinear, which is used for packing.
+            try:
+                del model
+                self.accelerator.empty_cache()
+                logger.info("Quantization is done, reloading model from saved directory...")
```
Comment (Kaihui-intel) [inline neural_compressor/torch/algorithms/weight_only/autoround.py]:
```suggestion
                logger.info(f"Quantization is done, reloading model from saved directory({self.output_dir})...")
```

File: neural_compressor/torch/algorithms/weight_only/autoround.py (hunk 2 above)
Comment (xin3he) [reply neural_compressor/torch/algorithms/weight_only/autoround.py:383]: I'm not sure what will happen if GGUF is set here, so I keep it here to ensure code robust.

File: neural_compressor/torch/algorithms/weight_only/autoround.py (hunk 4 above)
Comment (xin3he) [inline neural_compressor/torch/algorithms/weight_only/autoround.py]:
```suggestion
```

**Comment by xin3he** [comment]
> Overall, LGTM. It would be better not to mix example changes and new features in one PR.”

Right, most issues are found during enabling example, and those are mixed aiming quick development.

File: examples/pytorch/nlp/huggingface_models/language-modeling/quantization/auto_round/llama3/README.md [hunk 6]
```diff
@@ -0,0 +1,197 @@
...
+CUDA_VISIBLE_DEVICES=0 bash run_quant.sh --topology=Llama-3.1-8B --dtype=mxfp4_mixed --input_model=/models/Meta-Llama-3.1-8B-Instruct --output_model=Llama-3.1-8B-MXFP4-MXFP8
+```
+
+#### Llama 3.3 70B MXFP8
+
+```bash
+CUDA_VISIBLE_DEVICES=0 bash run_quant.sh --topology=Llama-3.3-70B --dtype=mxfp8 --input_model=/models/Llama-3.3-70B-Instruct/ --output_model=Llama-3.3-70B-MXFP8
+```
+
+#### Llama 3.3 70B MXFP4 (Mixed with MXFP8, Target_bits=5.8)
```
Comment (thuang6) [inline examples/pytorch/nlp/huggingface_models/language-modeling/quantization/auto_round/llama3/README.md:134]: besides giving exact target bits (here 5.8), is it possible to add an example cml to generate mixed bits model with INC autotune feature, for a new model, user may be able to get a good model throught autotune without trying different target-bits?

Comment (xin3he) [reply examples/pytorch/nlp/huggingface_models/language-modeling/quantization/auto_round/llama3/README.md:134]: Thanks, I updated the readme to  add an example cmd of llama3.1 8b autotune usage.  BTW, user still need to set the target bits list to avoid time-consuming autotune.

File: examples/pytorch/nlp/huggingface_models/language-modeling/quantization/auto_round/llama3/run_quant.sh [hunk 7]
```diff
@@ -0,0 +1,167 @@
...
+                ;;
+            *)
+                echo "Error: Unsupported dtype '$DTYPE' for topology '$TOPOLOGY'"
+                echo "Supported dtypes for Llama-3.1-8B: mxfp8, mxfp4_mixed"
+                exit 1
+                ;;
+        esac
+        ;;
+    "Llama-3.3-70B")
+        case "$DTYPE" in
```
Comment (chensuyue) [inline examples/pytorch/nlp/huggingface_models/language-modeling/quantization/auto_round/llama3/run_quant.sh:95]: Please support "MXFP4(pure)" as well for this model

File: examples/pytorch/nlp/huggingface_models/language-modeling/quantization/auto_round/llama3/run_quant.sh [hunk 8]
```diff
@@ -0,0 +1,167 @@
...
+echo "  Topology: $TOPOLOGY"
+echo "  Data Type: $DTYPE"
+echo "  Input Model: $INPUT_MODEL"
+echo "  Output Model: $OUTPUT_MODEL"
+
+# Set common parameters
+COMMON_ARGS="--quantize --enable_torch_compile --low_gpu_mem_usage --export_format auto_round"
+
+case "$TOPOLOGY" in
+    "Llama-3.1-8B")
```
Comment (chensuyue) [inline examples/pytorch/nlp/huggingface_models/language-modeling/quantization/auto_round/llama3/run_quant.sh:49]: Please support "MXFP4(pure)" as well for this model

**Review by mengniwang95 — APPROVED** [review]

File: examples/pytorch/nlp/huggingface_models/language-modeling/quantization/auto_round/llama3/README.md [hunk 9]
```diff
@@ -0,0 +1,236 @@
...
+```
+
+#### Llama 3.1 70B MXFP8
+
+RTN (Round-to-Nearest) is enough to keep accuracy.
+
+```bash
+CUDA_VISIBLE_DEVICES=0 bash run_quant.sh --topology=Llama-3.1-70B --dtype=mxfp8 --input_model=/models/Llama-3.1-70B-Instruct/ --output_model=Llama-3.1-70B-MXFP8
+```
+#### Llama 3.1 70B uNVFP4
```
Comment (thuang6) [inline examples/pytorch/nlp/huggingface_models/language-modeling/quantization/auto_round/llama3/README.md:158]: pls add NVFP4 recipe command line as well, though uNVFP4 is prioritized, NVFP4 is still target and someone may need

**Comment by chensuyue** [comment]
The CI issue looks not related to this PR, I will fix it in another PR.
